### PR TITLE
Group metrics by label

### PR DESCRIPTION
**Pull request by olix0r**
Previously, we maintained a map of labels for each metric. Because the same keys are used in multiple scopes, this causes redundant hashing & map lookup when updating metrics.

With this change, there is now only one map per unique label scope and all of the metrics for each scope are stored in the value. This makes metrics inserting faster and prepares for eviction of idle metrics.

The `Metric` type has been split into `Metric`, which now only holds metric metadata and is responsible for printing a given metric, and `Scopes` which holds groupings of metrics by label.

The `metrics!` macro is provided to make it easy to define `Metric` instances statically.

### CODE REVIEW

File: proxy/src/telemetry/metrics/counter.rs [hunk 1]
```diff
@@ -63,6 +63,8 @@ impl ops::AddAssign<Self> for Counter {
 }
 
 impl FmtMetric for Counter {
+    const KIND: &'static str = "counter";
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/counter.rs]: thanks for making this a const!

File: proxy/src/telemetry/metrics/http.rs [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+
+#[derive(Debug, Default)]
+pub(super) struct RequestMetrics {
+    total: Counter,
+}
+
+pub(super) type ResponseScopes = Scopes<ResponseLabels, ResponseMetrics>;
+
+#[derive(Debug, Default)]
+pub struct ResponseMetrics {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/http.rs]: 🙌

File: proxy/src/telemetry/metrics/http.rs [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+        Self::request_total.fmt_scopes(f, &self, |s| &s.total)?;
+
+        Ok(())
+    }
+}
+
+// ===== impl RequestMetrics =====
+
+impl RequestMetrics {
+    pub fn end(&mut self) {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/http.rs]: I like that the increment actions are now grouped in semantically related methods. Makes `Record` much less wordy! :)

File: proxy/src/telemetry/metrics/mod.rs [hunk 4]
```diff
@@ -63,6 +79,9 @@ pub use self::serve::Serve;
 /// differences in formatting each type of metric. Specifically, `Histogram` formats a
 /// counter for each bucket, as well as a count and total sum.
 trait FmtMetric {
+    /// The metric type.
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/mod.rs]: Nit: this comment isn't super descriptive --- maybe it should note that this is part of the help message?

File: proxy/src/telemetry/metrics/mod.rs [hunk 5]
```diff
@@ -73,35 +92,25 @@ trait FmtMetric {
...
-struct Metrics {
-    request_total: Metric<Counter, RequestLabels>,
-
-    response_total: Metric<Counter, ResponseLabels>,
-    response_latency: Metric<Histogram<latency::Ms>, ResponseLabels>,
-
-    tcp: TcpMetrics,
-
-    start_time: u64,
+struct Metric<'a, M: FmtMetric> {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/mod.rs]: Would be nice if there was a comment explaining how this is used.

File: proxy/src/telemetry/metrics/mod.rs [hunk 6]
```diff
@@ -73,35 +92,25 @@ trait FmtMetric {
...
+    start_time: Gauge,
 }
 
-#[derive(Debug, Clone)]
-struct Metric<M, L: Hash + Eq> {
-    name: &'static str,
-    help: &'static str,
-    values: IndexMap<L, M>
+#[derive(Debug)]
+struct Scopes<L: Display + Hash + Eq, M> {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/mod.rs]: Nit: again, would be nice to document what this is/what it's for?

File: proxy/src/telemetry/metrics/mod.rs [hunk 7]
```diff
@@ -111,255 +120,97 @@ struct Metric<M, L: Hash + Eq> {
...
-            help = self.help,
-        )?;
-
-        for (labels, histogram) in &self.values {
-            histogram.fmt_metric_labeled(f, self.name, labels)?;
-        }
+// ===== impl Scopes =====
 
-        Ok(())
+impl<L: Display + Hash + Eq, M> Default for Scopes<L, M> {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/mod.rs]: I think this can be derived?

File: proxy/src/telemetry/metrics/mod.rs [hunk 8]
```diff
@@ -111,255 +120,97 @@ struct Metric<M, L: Hash + Eq> {
...
+            start_time: t0.into(),
+            .. Root::default()
+        }
     }
 
-    fn read_bytes_total(&mut self, labels: TransportLabels) -> &mut Counter {
-        self.read_bytes_total.values
-            .entry(labels)
-            .or_insert_with(Counter::default)
+    fn request(&mut self, labels: RequestLabels) -> &mut http::RequestMetrics {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/mod.rs]: I wonder if the `metrics!` macro could generate these repetitive fns as well?

File: proxy/src/telemetry/metrics/http.rs [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+}
+
+impl fmt::Display for RequestScopes {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        if self.scopes.is_empty() {
+            return Ok(());
+        }
+
+        Self::request_total.fmt_help(f)?;
+        Self::request_total.fmt_scopes(f, &self, |s| &s.total)?;
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/http.rs]: I strongly suspect the accessor could be part of the generated impls or something so we don't have to pass a closure. However, the perf impact of a simple unboxed, non-capturing `Fn` closure like this is minimal so it doesn't really matter.

File: proxy/src/telemetry/metrics/mod.rs [hunk 10]
```diff
@@ -29,20 +29,36 @@
...
 use std::sync::{Arc, Mutex};
-use std::time;
+use std::time::{UNIX_EPOCH, Duration};
 
 use indexmap::IndexMap;
 
 use ctx;
 
+macro_rules! metrics {
+    { $( $name:ident : $kind:ty { $help:expr } ),+ } => {
```
Comment (hawkw) [inline proxy/src/telemetry/metrics/mod.rs]: This is really cool :D

**Comment by hawkw** [comment]
Purely out of curiosity, I'd love to see before/after benchmarks?

**Comment by olix0r** [comment]
@hawkw 

this branch:
```
running 3 tests
test record_many_dsts        ... bench:     104,570 ns/iter (+/- 14,207)
test record_one_conn_request ... bench:       1,164 ns/iter (+/- 256)
test record_response_end     ... bench:         397 ns/iter (+/- 85)
```

master:
```
test record_many_dsts        ... bench:     152,559 ns/iter (+/- 21,747)
test record_one_conn_request ... bench:       1,655 ns/iter (+/- 248)
test record_response_end     ... bench:         688 ns/iter (+/- 86)
```

File: proxy/src/telemetry/metrics/mod.rs (hunk 7 above)
Comment (hawkw) [reply proxy/src/telemetry/metrics/mod.rs]: Ah, good point --- nevermind then!

**Comment by hawkw** [comment]
@olix0r awesome :blazingly_fast:

**Pull request merged**